### PR TITLE
storage/engine: bump the memtable budget from 32 MB to 128 MB

### PR DESCRIPTION
**Pull request by petermattis**
This should alleviate write stalls when deleting a range.

See #9663

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9668)

### CODE REVIEW

**Comment by petermattis** [comment]
Cc @cockroachdb/stability

File: storage/engine/rocksdb.go [hunk 1]
```diff
@@ -261,15 +258,14 @@ func (c RocksDBCache) Release() {
...
-	memtableBudget int64              // Memory to use for the memory table.
-	maxSize        int64              // Used for calculating rebalancing and free space.
-	maxOpenFiles   int                // The maximum number of open files this instance will use.
-	stopper        *stop.Stopper
-	deallocated    chan struct{} // Closed when the underlying handle is deallocated.
+	rdb          *C.DBEngine
+	attrs        roachpb.Attributes // Attributes for this engine
+	dir          string             // The data directory
+	cache        RocksDBCache       // Shared cache.
+	maxSize      int64              // Used for calculating rebalancing and free space.
```
Comment (tschottdorf) [inline storage/engine/rocksdb.go]: used for more now.

File: storage/engine/rocksdb/db.cc [hunk 2]
```diff
@@ -1418,16 +1418,34 @@ DBStatus DBOpen(DBEngine **db, DBSlice dir, DBOptions db_opts) {
...
+  // uncompressed nature of the MemTable vs SSTables.
+  options.write_buffer_size = 8 << 20; // 8 MB
+  // How much memory should be allotted to memtables? Note that this
+  // is a peak setting, steady state should be lower. We set this
+  // relatively high to account for bursts of writes (e.g. due to a
+  // range deletion). In particular, we want this to be somewhat
+  // larger than than typical range size so that deletion of a range
+  // does not cause write stalls.
+  const uint64_t memtable_budget = 128 << 20; // 128 MB
+  options.max_write_buffer_number = memtable_budget / options.write_buffer_size;
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: std::max(<the above>, 2)?

File: storage/engine/rocksdb/db.cc [hunk 3]
```diff
@@ -1418,16 +1418,34 @@ DBStatus DBOpen(DBEngine **db, DBSlice dir, DBOptions db_opts) {
...
+  options.level0_file_num_compaction_trigger = 1;
```
Comment (tschottdorf) [inline storage/engine/rocksdb/db.cc]: Mind adding a comment for each?

**Comment by tamird** [comment]
Should the first commit remove `storage/engine/rocksdb/options_builder.{cc,h}`? Also, the grammar in that commit message seems to be broken -- we're not moving away from 4.9, since it's not in use in any version of CR.

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-:-KT9o1y4T3J-Vku-7EEb:b-dqse3d)*

**Comment by petermattis** [comment]
`storage/engine/rocksdb/options_builder.{cc,h}` doesn't exist as of the first commit. I'll fix the grammar shortly.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-:-KT9poS-DkRgcRqYibal:bahlbmd)*

**Comment by tamird** [comment]
Ah, this patch is exclusive of 4.9. Got it.

---

Reviewed 6 of 7 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-:-KT9rGzmalpKhNoAULBk:b-li245z)*

**Comment by petermattis** [comment]
Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 265 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-KT7KWDLBeyGIADX_mx7-r4-265:-KT9x1hLY_0z_imgIlRs:b95y5vb) ([raw file](https://github.com/cockroachdb/cockroach/blob/36697aee6c0f50ebe2b6236ea66db43679e62127/storage/engine/rocksdb.go#L265)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> used for more now.
</details>
What would you have me change the comment to?

---

*[storage/engine/rocksdb/db.cc, line 1432 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-KT7KWDKMB3xQqBMouI3-r4-1432:-KT9wClWN9a7R7WladhF:bym53g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/36697aee6c0f50ebe2b6236ea66db43679e62127/storage/engine/rocksdb/db.cc#L1432)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Kind of tricky that this is fixed but the range size can be changed by the user. Could you commit some of your experiments that "simulate" a DeleteRange of a range of x MB so that we can see how large a range size this covers typically?
</details>
I accidentally deleted that experiment. Note that I tried deleting up to a 100 MB batch and that worked without problem. The issue seems to be with multiple concurrent batches that add up to this value. 

---

*[storage/engine/rocksdb/db.cc, line 1433 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-KT7KWDKMB3xQqBMouI3-r4-1433:-KT9wS7-PLKHvLpJwnhi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36697aee6c0f50ebe2b6236ea66db43679e62127/storage/engine/rocksdb/db.cc#L1433)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `std::max(the above, 2)`?
</details>
Done.

---

*[storage/engine/rocksdb/db.cc, line 1434 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-KT7KWDKMB3xQqBMouI3-r4-1434:-KT9wT0ppWnBVb1taH8F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36697aee6c0f50ebe2b6236ea66db43679e62127/storage/engine/rocksdb/db.cc#L1434)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Mind adding a comment for each?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668)*

**Comment by tamird** [comment]
Reviewed 7 of 7 files at r10, 1 of 1 files at r11, 5 of 7 files at r12, 2 of 2 files at r13.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668)*

**Comment by tschottdorf** [comment]
Merging this as it appears that `omega` is really suffering. See https://github.com/cockroachdb/cockroach/issues/9749.

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 5 of 7 files at r12, 2 of 2 files at r13.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/engine/rocksdb.go, line 265 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-KT7KWDLBeyGIADX_mx7-r4-265:-KTLA4kShKDkTUo1xzoM:bzfky2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/36697aee6c0f50ebe2b6236ea66db43679e62127/storage/engine/rocksdb.go#L265)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What would you have me change the comment to?
</details>
I misread the diff, you didn't change anything about `maxSize`, correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-:-KTLABTg3FTctdKGZLj2:bnfp4nl)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/engine/rocksdb.go, line 265 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668#-KT7KWDLBeyGIADX_mx7-r4-265:-KTLBLCE9f5cRv0pxlgO:by6h7z3) ([raw file](https://github.com/cockroachdb/cockroach/blob/36697aee6c0f50ebe2b6236ea66db43679e62127/storage/engine/rocksdb.go#L265)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I misread the diff, you didn't change anything about `maxSize`, correct?
</details>
Nope. I removed the `memtableBudget` field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9668)*

**Pull request merged**